client/src/components/Workflow/Editor/Forms/FormTool.vue

```diff
@@ -163,9 +163,16 @@ export default {
         onLabel(newLabel) {
             this.$emit("onLabel", this.stepId, newLabel);
         },
+        /**
+         * Change event is triggered on component creation and input changes.
+         * @param { Object } values contains flat key-value pairs `prefixed-name=value`
+         */
         onChange(values) {
             this.mainValues = values;
-            this.postChanges();
+            if (this.lastId === this.id) {
```

Comment:
I think that is quite fragile (step ids can change), but you shouldn't need to fire on create at all; the state is always fresh from the server when it comes in from the props. Is firing on change only something the form system can handle ?

Reply:
The `id` in this case consists of `stepId` and `contentId`, the reason for the `onChange` handler to fire is that the `ToolForm` processes the raw tool model which is a nested dictionary containing the input fields etc. This tool model is traversed and the input values are collected and potentially formatted (e.g. numeric values) and then result in a flat dictionary containing prefixed key-value pairs i.e. the tool state.

Reply:
> The `id` in this case consists of `stepId` and `contentId`,

Right, that can change in a workflow, that's why I think this is fragile.



> This nested tool model is traversed and the input values are collected and potentially formatted (e.g. numeric values) and then result in a flat dictionary containing prefixed key-value pairs i.e. the tool state.

Why does the tool state contain formatted values ? It should not, right ?

Reply:
The tool state which the client submits to the backend in case values changed is a flat-dictionary which I think the backend does not provide but it can be generated from the form model.

Reply:
1) the values don't have to be flat, it's actually a lot easier if they're not
2) you mentioned formatted values in the tool state, why ?
3) I am still not understanding how flat or non-flat views of the (hopefully) same data require that an event is fired, if the value hasn't changed

Reply:
The backend provides a tool model e.g. `{ input_name: { type: "text", optional: "false", value: "value" }, ... }` but consumes `{ input_name: "value", ... }`. Maybe I am misunderstanding this and we should talk in a zoom about it to get it right?